File: app/src/main/java/com/yalantis/ui/fragment/BaseMvpFragment.java
package com.yalantis.ui.fragment;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.yalantis.contract.BaseMvpPresenter;
import com.yalantis.contract.BaseMvpView;


public abstract class BaseMvpFragment<T extends BaseMvpPresenter> extends BaseFragment implements BaseMvpView {

    protected T mPresenter;

    @Nullable
    @Override
    public final View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);
        mPresenter = getPresenterInstance();
        mPresenter.attachView(this);
        return view;
    }

    protected @NonNull abstract T getPresenterInstance();

    @Override
    public void onDestroyView() {
        mPresenter.detachView();
        super.onDestroyView();
    }

}

File: app/src/main/java/com/yalantis/base/BaseMvpFragment.kt
package com.yalantis.base

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup


abstract class BaseMvpFragment<in V: BaseMvpView, T : BaseMvpPresenter<V>> : BaseFragment(), BaseMvpView {

    abstract fun presenterInstance(): T

    protected var mPresenter: T = presenterInstance()

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = super.onCreateView(inflater, container, savedInstanceState)
        mPresenter.attachView(this as V)
        return view
    }

    override fun onDestroyView() {
        mPresenter.detachView()
        super.onDestroyView()
    }

}
